### ISSUE TITLE
librosa.util.normalize doesn't preserve dtype

### ISSUE THREAD

**Issue by carlthome** [librosa/librosa#580]
```python
In [1]: import librosa

In [2]: y, sr = librosa.load(librosa.util.example_audio_file())

In [3]: y.dtype
Out[3]: dtype('float32')

In [4]: y = librosa.util.normalize(y)

In [5]: y.dtype
Out[5]: dtype('float64')
```

**Comment by bmcfee** [comment]
We should add type preservation tests for this, as well as in `effects.*` and `istft`.

**Comment by jlw365** [comment]
I've started some work on this - it looks like in librosa.util.normalize the dtype preservation problem lies originally here,
https://github.com/librosa/librosa/blob/74dafbf497d552e77f861c543eee7aec1ff720ec/librosa/util/utils.py#L707

as `np.abs(S).dtype = float32`, and the following `.astype(np.float)` casts this to `float64`. Thus, when `length` is initialized using `mag` in lines 713, 716, 722, or 725, `length.dtype = float64`. 

It follows that the division of `Snorm = S / length` at lines 755, 751, or 757 is `float32/float64 = float64`, giving us a `Snorm.dtype = float64`.

Would you prefer to fix this in the initialization of `mag` (potentially via `mag = np.abs(S).astype(np.float32, casting='safe')`), or in the final division, via `Snorm = np.divide(S, length, dtype='float32')`?

Thanks for your help! Let me know if I should be looking in a different place/am on the wrong track/etc.

**Comment by carlthome** [comment]
IMO dtype should be preserved, but not forced to be float32. It should be enough to just get rid of `.astype(np.float)` as np.abs will preserve dtype as expected AFAIK.

**Comment by bmcfee** [comment]
@jlw365 thanks for taking this on!

> Would you prefer to fix this in the initialization of mag (potentially via mag = np.abs(S).astype(np.float32, casting='safe')), or in the final division, via Snorm = np.divide(S, length, dtype='float32')?

Any casting should happen as late as possible to avoid compounding overflow/underflow errors, so I like your suggestion of casting in the final division.

@carlthome I think it's going to be a bit trickier than you suggest, because not all norms are dtype-preserving.  For example, given a `float32` input, `norm='max'` will result in `length.dtype == np.float32`, but `norm=2` (or any p-norm) might inflate to `float64` due to the exponentiation: `float32 * float32 -> float64`.

**Comment by carlthome** [comment]
Oh, dang! I just assumed NumPy would respect an explicitly set dtype and maybe provide an underflow warning, instead of just doing an automatic copy behind the scenes. :cry: 

Maybe best to follow NumPy's convention then and maintain input dtype until some value needs more bits.

**Comment by bmcfee** [comment]
@jlw365 you still working on this one?

**Comment by jlw365** [comment]
Sorry - got a bit sidetracked on this one. Anyone should feel free to work on this issue - not sure I have the time at the moment.

**Comment by bmcfee** [comment]
@jlw365 no problem, thanks for the update.

**Comment by bmcfee** [comment]
Fixed by merging #636
